lib/utils/platform/iot_threads_afr.c

```diff
@@ -0,0 +1,350 @@
+    }
+    else
+    {
+         ( void ) xSemaphoreGive( ( SemaphoreHandle_t ) & internalMutex->xMutex );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+bool IotSemaphore_Create( IotSemaphore_t * const pNewSemaphore,
+                             uint32_t initialValue,
+                             uint32_t maxValue )
+{
+    _iot_sem_internal_t * internalSemaphore = ( _iot_sem_internal_t * ) pNewSemaphore;
+    IotLogDebug( "Creating new semaphore %p.", pNewSemaphore );
+
+    ( void ) xSemaphoreCreateCountingStatic( maxValue, initialValue, &internalSemaphore->xSemaphore );
+
+    return true;
+}
+
+/*-----------------------------------------------------------*/
+
+uint32_t IotSemaphore_GetCount( IotSemaphore_t * const pSemaphore )
+{
+    _iot_sem_internal_t * internalSemaphore = ( _iot_sem_internal_t * ) pSemaphore;
+    UBaseType_t count = 0;
+
+    count = ( UBaseType_t ) uxSemaphoreGetCount( ( SemaphoreHandle_t ) &internalSemaphore->xSemaphore );
+
+    IotLogDebug( "Semaphore %p has count %d.", pSemaphore, count );
+
+    return ( uint32_t ) count;
+}
+
+/*-----------------------------------------------------------*/
+
+void IotSemaphore_Destroy( IotSemaphore_t * const pSemaphore )
+{
+    _iot_sem_internal_t * internalSemaphore = ( _iot_sem_internal_t * ) pSemaphore;
+
+    IotLogDebug( "Destroying semaphore %p.", internalSemaphore );
+
+    vSemaphoreDelete( ( SemaphoreHandle_t ) &internalSemaphore->xSemaphore );
+}
+
+/*-----------------------------------------------------------*/
+
+void IotSemaphore_Wait( IotSemaphore_t * pSemaphore )
+{
+    _iot_sem_internal_t * internalSemaphore = ( _iot_sem_internal_t * ) pSemaphore;
+
+    IotLogDebug( "Waiting on semaphore %p.", internalSemaphore );
+
+    /* Take the semaphore using the FreeRTOS API. */
+    if( xSemaphoreTake( ( SemaphoreHandle_t ) &internalSemaphore->xSemaphore,
+                          portMAX_DELAY ) != pdTRUE )
+    {
+        IotLogWarn( "Failed to wait on semaphore %p.",
+                       pSemaphore);
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+bool IotSemaphore_TryWait( IotSemaphore_t * const pSemaphore )
+{
+    _iot_sem_internal_t * internalSemaphore = ( _iot_sem_internal_t * ) pSemaphore;
+
+    IotLogDebug( "Attempting to wait on semaphore %p.", internalSemaphore );
+
+    return IotSemaphore_TimedWait( pSemaphore, 0 );
+}
+
+/*-----------------------------------------------------------*/
+
+bool IotSemaphore_TimedWait( IotSemaphore_t * const pSemaphore,
+                                uint64_t timeoutMs )
+{
+    _iot_sem_internal_t * internalSemaphore = ( _iot_sem_internal_t * ) pSemaphore;
+
+    /* Take the semaphore using the FreeRTOS API. */
+    if( xSemaphoreTake( ( SemaphoreHandle_t ) &internalSemaphore->xSemaphore,
+                          timeoutMs * configTICK_RATE_HZ / 1000 ) != pdTRUE )
```

Comment:
Is there an overflow or underflow lurking around here? 

timeoutMs * configTICK_RATE_HZ / 1000

Reply:
Overflow if we change the time to be 32 bit, since it was 64-bit ... 

What do we want to do here? If we divide first it will round, tick rates are typically things like 100Hz which is 10ms, or possibly 1000 which is 1ms.